Add an API for clang dependency scanner to perform module lookup by name alone

ahatanaka opened the pull request:
Rebase onto `next` branch. The old pull request is here: https://github.com/apple/llvm-project/pull/3099.

ahatanaka committed:
[clang][clang-scan-deps] Add an experimental C API for returning the list of
module and file dependencies of a module for a particular compiler invocation

rdar://64538073

ahatanaka committed:
Address review comments

- Define a subclass of FrontendAction and call loadModule there.
- Remove the API for creating a worker just for
  clang_experimental_DependencyScannerWorkerByModName_getFileDependencies_v1.
- Add the capability to get the dependencies based on the module name to
  clang-scan-deps.

ahatanaka committed:
Address review comments

- Create a fake virtual source file instead of asking users to create
  one and pass it to the command line.
- Remove the "translation-units" output from test case
  modules-full-by-mod-name.cpp. The information isn't useful as the
  source file name isn't passed on the command line in this case.
- Inherit from `PreprocessOnlyAction`.

jansvoboda11 commented on clang/include/clang-c/Dependencies.h:
```diff
@@ -222,6 +222,17 @@ clang_experimental_DependencyScannerWorker_getFileDependencies_v1(
+ * but get the dependencies by module name alone.
+ */
+CINDEX_LINKAGE CXFileDependencies *
+clang_experimental_DependencyScannerWorkerByModName_getFileDependencies_v0(
```
Continuing our discussion on naming from the original PR.

Yes, the `CXFileDependencies` type we're returning here does contain both `FileDeps` and `ModuleDeps`. However, I think its name and the name of the existing functions (`getFileDependencies`) don't refer to the contents of the result, but rather to the fact that they will contain/compute dependencies of a C/ObjC **file**.

So if we're scanning dependencies of a named module, I think it's reasonable to call the new function `getModuleDependencies` and consider renaming `CXFileDependencies` to simply `CXDependencies`.

@Bigcheese do you have any thoughts on this?

Bigcheese replied:
I think the current naming we have is a bit confusing now. I think it would be better to have the following:

`CXDependencies`
`clang_experimental_DependencyScannerWorker_getDependenciesByModName_v0`
`clang_experimental_DependencyScannerWorker_getDependenciesByFile_v0`

Not sure if we should do this now as we'll need to update users, but it seems to make more sense.

jansvoboda11 commented on clang/include/clang/Frontend/FrontendActions.h:
```diff
@@ -299,6 +299,14 @@ class PrintPreprocessedAction : public PreprocessorFrontendAction {
   bool hasPCHSupport() const override { return true; }
 };
 
+class GetDependiciesByModuleNameAction : public PreprocessOnlyAction {
```
Typo. `GetDependiciesByModuleNameAction` -> `GetDependenciesByModuleNameAction`.

jansvoboda11 replied:
I think inheriting from `PreprocessOnlyAction` here is fine. 👍 We don't need to load any PCH files.

jansvoboda11 commented on clang/lib/Tooling/DependencyScanning/DependencyScanningWorker.cpp:
```diff
@@ -313,13 +332,30 @@ llvm::Error DependencyScanningWorker::computeDependencies(
     /// file system requests that are made by the driver do not go through the
     /// dependency scanning filesystem.
-    tooling::ClangTool Tool(CDB, Input, PCHContainerOps, RealFS, Files);
+    assert(!!ModuleName == Input.empty());
```
Do we use `!!` anywhere else? I think `!= nullptr` would be clearer.

jansvoboda11 commented on clang/test/ClangScanDeps/modules-full-by-mod-name.cpp:
```diff
@@ -0,0 +1,86 @@
+// RUN: sed -e "s|DIR|%/t.dir|g" %S/Inputs/modules_cdb_by_mod_name.json > %t.cdb
+// RUN: sed -e "s|DIR|%/t.dir|g" %S/Inputs/modules_cdb_clangcl_by_mod_name.json > %t_clangcl.cdb
+//
+// RUN: echo %t.dir > %t.result
+// RUN: clang-scan-deps -compilation-database %t.cdb -j 4 -format experimental-full \
+// RUN:   -mode preprocess-minimized-sources -module-name=header1 >> %t.result
+// RUN: cat %t.result | sed 's:\\\\\?:/:g' | FileCheck --check-prefixes=CHECK,CHECK-NO-ABS %s
```
This is identical to the test below except for the checked prefix. We don't need to test the logic related to path arguments. Can you delete these `RUN` lines and any `CHECK-NO-ABS-NOT` lines?

jansvoboda11 commented on clang/tools/c-index-test/core_main.cpp:
```diff
@@ -1011,6 +1090,19 @@ int indextest_core_main(int argc, const char **argv) {
 
+  if (options::Action == ActionType::ScanDepsByModName) {
+    if (options::InputFiles.empty()) {
+      errs() << "error: missing working directory\n";
```
The condition checks input files and error message reports working directory. Is that correct? If so, can you explain in a comment?

ahatanaka replied:
Yes, this is correct. When `-scan-deps-by-mod-name` or `-scan-deps` is passed, `InputFiles` contains the working directory name, which gets passed to `scanDepsByModName`. Perhaps we can rename it to `InputFilesOrDirectories` or something to make it less confusing.

jansvoboda11 replied:
Interesting. Can you put the explanation into a comment in the code?

jansvoboda11 commented on clang/tools/clang-scan-deps/ClangScanDeps.cpp:
```diff
@@ -194,6 +194,10 @@ llvm::cl::opt<bool> SkipExcludedPPRanges(
     llvm::cl::init(true), llvm::cl::cat(DependencyScannerCategory));
 
+llvm::cl::opt<std::string> ModuleName("module-name", llvm::cl::Optional,
+                                      llvm::cl::desc("module name"),
```
Add a more detailed description please.

jansvoboda11 commented on clang/tools/libclang/CDependencies.cpp:
```diff
@@ -220,14 +220,25 @@ getFileDependencies(CXDependencyScannerWorker W, int argc,
       Compilation.push_back(argv[i]);
-  else {
+    if (Worker->getModuleName())
+      Compilation.push_back(GetFakeSrcFile());
```
We're already doing that in `DependencyScanningWorker.cpp`, no? Is there a reason we need to inject the name of the fake source file here too?

ahatanaka replied:
I had to add the argument before `CompilerInvocation::CreateFromArgs`, which gets called in `newInvocation`, gets called. Is there a way to avoid adding the argument or other places where the argument can be added instead of adding it here?

jansvoboda11 replied:
And the motivation is that the driver/cc1 command-line parsers in always require an input file? If so, we should document it here.

You mean the call to `newInvocation` from `DependencyScanningWorker::computeDependenciesForClangInvocation`? Or in `ToolInvocation::run`?

I can imagine even the call to `createInvocationFromCommandLine` down on line 244 being problematic, what do you think?

ahatanaka replied:
I meant the call to `newInvocation` in `DependencyScanningWorker::computeDependenciesForClangInvocation`. The fake source file name has to be added before `CompilerInvocation::CreateFromArgs` is called because `-` will be used as the default input If there is no input file in the argument list (see `ParseFrontendArgs` in `CompilerInvocation.cpp`). 

You are right that the fake source file name is needed for `createInvocationFromCommandLine` too. Without the fake source file, clang will error out early in `Driver::BuildActions` with error message `no input files`.

jansvoboda11 replied:
Can you please rebase? With [this PR](https://github.com/apple/llvm-project/pull/3168) merged, none of these command-line modifications should be necessary. Doing the work once in `DependencyScanningWorker` should do it.

jansvoboda11 reviewed (changes requested):
Nice work, this is getting closer!

ahatanaka committed:
Address review comments

jansvoboda11 commented on clang/include/clang-c/Dependencies.h:
```diff
@@ -222,6 +222,17 @@ clang_experimental_DependencyScannerWorker_getFileDependencies_v1(
+ * but get the dependencies by module name alone.
+ */
+CINDEX_LINKAGE CXFileDependencies *
+clang_experimental_DependencyScannerWorker_getDependenciesByModName_v0(
```
Can we change `...ModName` to `...ModuleName`? I think being terse here might make the API harder to understand.

jansvoboda11 commented on clang/test/ClangScanDeps/Inputs/modules_cdb_by_mod_name.json:
```diff
@@ -0,0 +1,22 @@
+},
+{
+  "directory": "DIR",
+  "command": "clang -E -IInputs -fmodules -fcxx-modules -fmodules-cache-path=DIR/module-cache -fimplicit-modules -fimplicit-module-maps -o b.o -x c++",
```
Why are there so many compile commands in this file when the test only scans dependencies of a single module?

jansvoboda11 commented on clang/test/ClangScanDeps/Inputs/modules_cdb_clangcl_by_mod_name.json:
```diff
@@ -0,0 +1,22 @@
+},
+{
+  "directory": "DIR",
+  "command": "clang-cl /E /IInputs /clang:-fmodules /clang:-fcxx-modules /clang:-fmodules-cache-path=DIR/module-cache_clangcl /clang:-fimplicit-modules /clang:-fimplicit-module-maps -o b.o /clang:-x /clang:c++ --",
```
Ditto `clang/test/ClangScanDeps/Inputs/modules_cdb_by_mod_name.json`.

jansvoboda11 commented on clang/test/Index/Core/scan-deps-by-mod-name.m:
```diff
@@ -0,0 +1,34 @@
+// RUN: rm -rf %t.mcp
+// RUN: echo %S > %t.result
+// RUN: c-index-test core --scan-deps-by-mod-name -module-name=ModA %S -- %clang -c -I %S/Inputs/module \
+// RUN:     -fmodules -fmodules-cache-path=%t.mcp \
```
This `RUN` line only differs from the one above in `-fimplicit-module-maps`. AFAIK there's no difference how implicit module maps are handled between dependency scan for a file vs module name. Why do we need to test it here?

jansvoboda11 reviewed (changes requested):
I think the tests need a bit more work, they don't need to test unrelated functionality. Besides that, I only have a couple of small nits.

ahatanaka committed:
Address review comments

- Rebase.
- Rename functions.
- Remove unneeded test cases.

jansvoboda11 commented on clang/include/clang/Tooling/DependencyScanning/DependencyScanningWorker.h:
```diff
@@ -97,6 +97,11 @@ class DependencyScanningWorker {
 
   ScanningOutputFormat getFormat() const { return Format; }
 
+  const char *getModuleName() const { return ModuleName; }
```
This is not called anywhere.

jansvoboda11 commented on clang/include/clang/Tooling/DependencyScanning/DependencyScanningWorker.h:
```diff
@@ -113,6 +118,9 @@ class DependencyScanningWorker {
   llvm::IntrusiveRefCntPtr<FileManager> Files;
   ScanningOutputFormat Format;
+
+  const char *ModuleName = nullptr;
+  std::unique_ptr<llvm::MemoryBuffer> FakeMemBuffer;
```
Sorry I didn't catch this earlier.

Looking at the dependency scanner architecture again, I don't think these fields should live inside `DependencyScanningWorker`. They introduce a state that cannot be shared between multiple invocations of the worker. If you call `Worker->setModuleName()`, you'd need to remember to call `Worker->unsetModuleName()` to clear the temporary state.

The only place the module name is being used is in `DependencyScanningWorker::computeDependencies` where we're setting up `DependencyScanningAction`. Why don't we pass the optional module name as a parameter to the function?

The same goes for `FakeMemBuffer`. This can be just a variable inside `computeDependencies`, can't it?

ahatanaka replied:
Do we want to pass the module name as an optional argument to all the functions that indirectly call `computeDependencies`, e.g., `computeDependenciesForClangInvocation`, `getFullDependencies`, `getFileDependencies`?

`DependencyScanningTool` has member `ModuleName` too. Is it necessary to remove that too?

jansvoboda11 replied:
Yes, we'll need to thread the module name through all the functions I think.

Yes, the `DependencyScanningTool` instances are being reused for multiple dependency scans, so it's the same deal as `DependencyScanningWorker`.

jansvoboda11 commented on clang/include/clang/Frontend/FrontendActions.h:
```diff
@@ -299,6 +299,15 @@ class PrintPreprocessedAction : public PreprocessorFrontendAction {
 };
 
+class GetDependenciesByModuleNameAction : public PreprocessOnlyAction {
+  const char *ModuleName;
```
Can we use `StringRef` instead of `const char *` everywhere?

jansvoboda11 commented on clang/lib/Tooling/DependencyScanning/DependencyScanningWorker.cpp:
```diff
@@ -262,6 +279,8 @@ class DependencyScanningAction : public tooling::ToolAction {
   ExcludedPreprocessorDirectiveSkipMapping *PPSkipMappings;
   ScanningOutputFormat Format;
+  const char *ModuleName;
+  const llvm::MemoryBuffer *FakeMemBuffer;
```
I think saw Duncan refactoring some instances of `llvm::MemoryBuffer *` to `llvm::MemoryBufferRef` lately. Would it make sense to use that instead?

ahatanaka committed:
Address review comments

- Remove member `ModuleName` and thread it through functions. Remove
  member `FakeMemBuffer` too and make it a variable inside
  `computeDependencies`.
- Use `MemoryBufferRef` instead of `MemoryBuffer *`.
- Use `StringRef` instead of `const char *`.

jansvoboda11 commented on clang/include/clang/Tooling/DependencyScanning/DependencyScanningTool.h:
```diff
@@ -83,7 +83,7 @@ class DependencyScanningTool {
   llvm::Expected<std::string>
   getDependencyFile(const tooling::CompilationDatabase &Compilations,
-                    StringRef CWD);
+                    StringRef CWD, StringRef ModuleName);
```
Can you document what's the expected behavior of the function when `ModuleName` is not empty? For other functions that take this argument as well.

Do you think changing the argument type `Optional<StringRef> ModuleName = None` might make it more obvious for callers that this isn't something they always have to provide?

ahatanaka replied:
Yes, I think using `Optional` makes it more obvious that the parameter is optional.

jansvoboda11 reviewed (commented):
Left one last nit inline. Not sure how we usually treat PRs, but I think it'd be nice to squash the commits and give this PR more descriptive name.

ahatanaka committed:
Address review comments

- Use Optional<String> as the type of ModuleName.
- Add comments explaining what happens if the passed ModuleName isn't
  empty.

jansvoboda11 commented:
LGTM, thanks for working through this Akira! @Bigcheese, @hyp, @dexonsmith do you have any additional feedback?
